### DISCUSSION TITLE
Windows pre-boot screen splits to multiple displays after EFI Sanboot

### DISCUSSION THREAD

**Discussion by darksidemilk** [ipxe/ipxe#1253] [General]
So I use [FOGProject ](https://github.com/FOGProject/fogproject/tree/working-1.6) which uses the latest version of ipxe.
We have sanboot working in EFI as described in this commit
https://github.com/ipxe/ipxe/commit/b1c13cc43
(which is amazing and thank you for getting that implemented by the way) 

So overall this is working but I've noticed an issue on some physical devices (doesn't seem to happen on a virtual machine).
When we 'boot to hard disk' from the fog ipxe boot menu which uses the command 

```
sanboot --no-describe --drive 0x80 || sanboot --no-describe --drive 0x81 || sanboot --no-describe --drive 0x82
```

It will boot to Windows but creates multiple displays (how many varies by device/monitor).

For example, it might show a single Windows logo normally and then jumps to multiple spinning logos

![20240712_081025](https://github.com/user-attachments/assets/a85c2896-4cb6-4670-b5cc-66a9c354e401)F

For the most part this isn't a huge issue as it returns to normal once Windows loads. 

However, in the case of imaging a Windows machine and utilizing Windows sysprep, the specialize phase runs the initial setup in that screen and the duplicated display issue exists there. If something were to go wrong during this part, debugging from the machine would be very difficult as I can't quite read the output here...

![image](https://github.com/user-attachments/assets/d3fed363-02b0-48e8-8860-359ae9e6d53b)

I have tried switching from ipxe.efi to snponly.efi and get the same thing. 

I've seen weird issues with the logo in the initial loading screen when using different boot environments to get into windows, but it's usually just the logo, this is a new one.

We previously chained a refined.efi boot file to boot to disk using that utility and it didn't have this issue but sanboot direct to disk is much simpler.

**Comment by NiKiZe** [comment]
I don't see how this could even be debugged without more info, so moving to discussion until such information has come to light.

First of, let's get rid of external factors, can you reproduce this with snponly.efi from boot.ipxe.org, pressing ctrl+b to get to the shell without network and then sanboot from there?
Could you also do this test without but booting iPXE from usb device, so network available?

Could you list which devices this happens with, manufacturer, model, and bios version, also which graphic cards these devices have.
As well as which version of Windows.

All this to narrow down what the trigger is.

Do you have any graphic drivers loaded, slipped in to that stage of the process?

**Reply by darksidemilk** [reply]
Seeing this with Windows 10 22H2 and Windows 11 23H2

Some devices I've seen this on

* Lenovo L390 Yoga | bios - R10ET60W (1.45) | GPU - intel UHD 620 |  Win 11 23H2 | (laptop in picture)
* Lenovo L13 Yoga | bios - R15ET60W (1.41 ) | GPU - intel UHD Graphics | Win 11 23H2 | (has same issue as picture)
* HP Z2 G9 Mini | bios -U50 Ver. 03.02.03 | GPU - Nvidia T400 4GB | Win 10 22H2 | (splits into 2 screens)
* VMware Workstation 17 VM | Win 11 23H2 | Does not occur

I'll make a USB bootable pxe drive and give it a go.

As far as loaded graphics drivers, we're just in a pxe menu on the fog side, we haven't loaded the fog linux kernels or inits at that point. Then in the windows side not doing anything with drivers at that point as far as I know. In the picture with the sysprep script running, that is installing injected drivers but the issue still occurs on the next boot.

**Reply by NiKiZe** [reply]
Can you check if the HP is dual intel+Nvidia?
Do you have any graphics in iPXE, background image or similar?

I would like to know if this is reproducible with wimboot and default boot.wim? To make the test case as minimal as possible.

**Reply by darksidemilk** [reply]
I'm having a little trouble getting the ipxe USB bootable. I tried doing a dd inside windows of the ipxe.usb on boot.ipxe.org but the USB wasn't bootable. I don't have access to a physical linux machine at the moment. 
I was however able to put the snponly.efi on the efi partition of one of the machines and adjusted the bcd settings. 

I got the ctrl+B prompt and ran

```
sanboot --drive 0 --no-describe
```

it booted without the graphical anomaly. 

I tried doing the same thing with my networked ipxe.efi file (as in I put it on the EFI partition of the machine), I didn't have the Ctrl+B option there, so I did a ctrl+c when I got to the menu and still had the problem.
Then I did it again, but this time I hit ctrl+c before the network configured and downloaded the network boot files to get to the menu. When I did that and got to an ipxe command line it booted without issue.

All these tests were on the lenovo L390 yoga

**Reply by darksidemilk** [reply]
> Can you check if the HP is dual intel+Nvidia? Do you have any graphics in iPXE, background image or similar?
> 
> I would like to know if this is reproducible with wimboot and default boot.wim? To make the test case as minimal as possible.

The HP has integrated video disabled, 100% sure.
We do have a custom ipxe background image. The problem doesn't happen if sanboot happens before that image loads.
The image itself loads without issue under normal operations.

I'm not familiar with wimboot, is that like a wds thing?

**Reply by darksidemilk** [reply]
Now you got me thinking it could be the background image, so I removed the background file from the network instance to test.

i.e. in this case on the fog server that is hosting ipxe I ran this to rename the background images
```
mv /var/www/html/fog/service/ipxe/bg.png bg.old
mv /var/www/html/fog/service/ipxe/bgdark.png bgdark.old
```

Then I booted to pxe normally and the anamoly didn't occur.

So I'm finding that running sanboot to boot into windows on an efi system where a background image is used in the pxe menu a graphical anamoly occurs where the windows pre-execution/loading screen is duplicated on the monitor.

**Reply by NiKiZe** [reply]
I would try playing with changing resolution in iPXE before the sanboot command, and for now assume that windows isn't playing nice with the graphics mode set in iPXE.

Wimboot used to boot winpe https://ipxe.org/wimboot

To boot from usb in efi mode, grab the efi bootloader and place it as bootx64.efi on a fat32 partition, no BCD or other modifications, but I don't think any more of that testing will be needed.

**Reply by darksidemilk** [reply]
you can change the resolution in ipxe via ipxe command line?

**Reply by darksidemilk** [reply]
https://ipxe.org/cmd/console

This is what I should mess with?

**Reply by darksidemilk** [reply]
Adjusting the sanboot commands to be prefixed with a `console` like `console && sanboot...`  seems to have fixed the issue. 
It resets the console to the display resolution and then loads into windows normal.

Thank you @NiKiZe for the help! I thought this was going to be much more complicated.

**Reply by darksidemilk** [reply]
Just for reference, before I tried adding the `console` command to reset the resolution, I tried different images and different sizes like 800 x 600 but that didn't make a difference. Seemed to be happening when any image was being loaded. We probably have a safe resolution being set in the Fog code for the pxe menu.

**Comment by eoli3n** [comment]
Same issue here, let's test the console fix
